flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/PartitionTransformUdf.java

```diff
@@ -0,0 +1,159 @@
+      Transform<String, String> truncate = Transforms.truncate(Types.StringType.get(), num);
+      return truncate.apply(value);
+    }
+
+    public byte[] eval(int num, byte[] value) {
+      ByteBuffer wrap = ByteBuffer.wrap(value);
+      Transform<ByteBuffer, ByteBuffer> truncate = Transforms.truncate(Types.BinaryType.get(), num);
+      ByteBuffer byteBuffer = truncate.apply(wrap);
+      byte[] out = new byte[byteBuffer.remaining()];
+      byteBuffer.get(out, 0, out.length);
+      return out;
+    }
+
+    public BigDecimal eval(int num, BigDecimal value) {
+      Transform<BigDecimal, BigDecimal> truncate = Transforms.truncate(
+          Types.DecimalType.of(value.precision(), value.scale()),
+          num);
+      return truncate.apply(value);
+    }
+
+    @Override
+    public TypeInference getTypeInference(DataTypeFactory typeFactory) {
+      return TypeInference.newBuilder()
+          .outputTypeStrategy(callContext -> {
+            LogicalTypeRoot typeRoot = callContext.getArgumentDataTypes().get(1).getLogicalType().getTypeRoot();
+            if (typeRoot == LogicalTypeRoot.CHAR) {
+              return Optional.of(DataTypes.STRING());
+            } else if (typeRoot == LogicalTypeRoot.BINARY) {
+              return Optional.of(DataTypes.BYTES());
+            }
+            return Optional.of(callContext.getArgumentDataTypes().get(1));
+          })
+          .build();
+    }
+  }
+
+  public static class Bucket extends ScalarFunction {
+
+    public int eval(int num, int value) {
+      Transform<Object, Integer> bucket = Transforms.bucket(Types.IntegerType.get(), num);
+      return bucket.apply(value);
+    }
+
+    public int eval(int num, long value) {
+      Transform<Object, Integer> bucket = Transforms.bucket(Types.LongType.get(), num);
+      return bucket.apply(value);
+    }
+
+    public int eval(int num, BigDecimal value) {
+      Transform<Object, Integer> bucket =
+          Transforms.bucket(Types.DecimalType.of(value.precision(), value.scale()), num);
+      return bucket.apply(value);
+    }
+
+    public int eval(int num, String value) {
+      Transform<Object, Integer> bucket = Transforms.bucket(Types.StringType.get(), num);
+      return bucket.apply(value);
+    }
+
+    public int eval(int num, byte[] value) {
+      ByteBuffer wrap = ByteBuffer.wrap(value);
+      Transform<Object, Integer> bucket = Transforms.bucket(Types.BinaryType.get(), num);
+      return bucket.apply(wrap);
+    }
+
+    public int eval(int num, @DataTypeHint("DATE") LocalDate value) {
+      Transform<Object, Integer> bucket = Transforms.bucket(Types.DateType.get(), num);
+      int days = (int) value.toEpochDay();
+      return bucket.apply(days);
+    }
+
+    public int eval(int num, @DataTypeHint("TIME") LocalTime value) {
+      Types.TimeType type = Types.TimeType.get();
+      Transform<Object, Integer> bucket = Transforms.bucket(type, num);
+      long micros = TimeUnit.NANOSECONDS.toMicros(value.toNanoOfDay());
```

Comment:
Can you use the conversion functions in `DateTimeUtil`?

Reply:
done